### PR TITLE
Add Diffie-Hellman key exchanges

### PR DESCRIPTION
**Pull request by grbean**
Signed-off-by: Greg Bean <greg.bean@intel.com>

### CODE REVIEW

File: protocol/src/main/java/org/fidoalliance/fdo/protocol/CryptoService.java [hunk 1]
```diff
@@ -1206,6 +1208,23 @@ public KeyExchangeResult getSharedSecret(byte[] message, Composite ownState, Key
       case Const.ASYMKEX2048_ALG_NAME:
       case Const.ASYMKEX3072_ALG_NAME:
         return getAsymkexSharedSecret(message, ownState, decryptionKey);
+      case DiffieHellman.DH14_ALG_NAME:
+      case DiffieHellman.DH15_ALG_NAME:
+        // All owner states have our message as element 0: restore our keyExchange
+        // from the KE state, which is element 2 in the owner state.
```
Comment (johneasterday) [inline protocol/src/main/java/org/fidoalliance/fdo/protocol/CryptoService.java:1214]: very picky here - comment is correct but was confusing at first. Adding "(THIRD_KEY)" after "element 2" would make it clearer.

Comment (grbean) [reply protocol/src/main/java/org/fidoalliance/fdo/protocol/CryptoService.java]: Fixed.

**Review by johneasterday — APPROVED** [review]

**Review by DukeDavis12 — APPROVED** [review]

File: protocol/src/main/java/org/fidoalliance/fdo/protocol/CryptoService.java [hunk 2]
```diff
@@ -1206,6 +1208,26 @@ public KeyExchangeResult getSharedSecret(byte[] message, Composite ownState, Key
...
+      case DiffieHellman.DH14_ALG_NAME:
+      case DiffieHellman.DH15_ALG_NAME:
+        // All owner states have our message as element 0 (FIRST_KEY) and the algorithm name
+        // in element 1 (SECOND_KEY).
+        //
+        // Restore our keyExchange from the KE state,
+        // which is element 2 (THIRD_KEY) in the owner state.
+        DiffieHellman.KeyExchange ke = new KeyExchange(ownState.getAsComposite(Const.THIRD_KEY));
+        try {
+          return new KeyExchangeResult(
```
Comment (vikashmishra07) [inline protocol/src/main/java/org/fidoalliance/fdo/protocol/CryptoService.java:1220]: Where is the catch block for the above statement?

File: protocol/src/main/java/org/fidoalliance/fdo/protocol/CryptoService.java [hunk 3]
```diff
@@ -1234,6 +1256,23 @@ public Composite getKeyExchangeMessage(String kexSuiteName, String party, Key ow
       case Const.ASYMKEX3072_ALG_NAME:
         return getAsymkexMessage(
             Const.ASYMKEX3072_ALG_NAME, Const.ASYMKEX3072_RANDOM_SIZE, party, ownerKey);
+      case DiffieHellman.DH14_ALG_NAME:
+      case DiffieHellman.DH15_ALG_NAME:
+        DiffieHellman.KeyExchange ke = DiffieHellman.buildKeyExchange(kexSuiteName);
+        try {
+          return Composite.newArray()
```
Comment (vikashmishra07) [inline protocol/src/main/java/org/fidoalliance/fdo/protocol/CryptoService.java:1263]: Where is the catch block for the above statement?

File: protocol/src/main/java/org/fidoalliance/fdo/protocol/CryptoService.java (hunk 2 above)
Comment (grbean) [reply protocol/src/main/java/org/fidoalliance/fdo/protocol/CryptoService.java:1220]: There is no catch block, it is a try-finally guarding the 'ke' resource.  Try-finally guarantees that the finally clause will execute no matter how the block is exited.

File: protocol/src/main/java/org/fidoalliance/fdo/protocol/CryptoService.java [hunk 4]
```diff
@@ -1206,6 +1208,26 @@ public KeyExchangeResult getSharedSecret(byte[] message, Composite ownState, Key
...
+        try {
+          return new KeyExchangeResult(
+              ke.computeSharedSecret(new BigInteger(1, message)).toByteArray(), new byte[0]);
+        } finally {
+          try {
+            ke.destroy();
+          } catch (DestroyFailedException e) {
+            // this should never happen
+            assert false;
+            throw new RuntimeException(e);
```
[the remark this answers is not in the document]
Comment (grbean) [reply protocol/src/main/java/org/fidoalliance/fdo/protocol/CryptoService.java:1228]: To reinforce the comment on the previous line, that this should never happen.

**Pull request merged**